storage/rocksdb/ha_rocksdb.cc

```diff
@@ -3072,6 +3093,82 @@ class Rdb_snapshot_status : public Rdb_tx_list_walker {
           tx->get_delete_count());
     }
   }
+
+  void populateDeadlockBuffer() {
+    uint max_key_size = 16;
+    m_data += "---------\n"
+              "DEADLOCKS\n"
+              "---------\n"
+              "LIST OF DEADLOCKS FOR EACH SESSION:\n";
+    std::vector<rocksdb::DeadlockPath> dlock_buffer =
+        rdb->GetDeadlockInfoBuffer();
+    for (auto path_entry : dlock_buffer) {
+      if (!path_entry.limit_exceeded) {
+        m_data += "\n*** DEADLOCK PATH\n"
+                  "=========================================\n";
+        for (auto txn : path_entry.path) {
+          GL_INDEX_ID gl_index_id = {
+              .cf_id = txn.m_cf_id,
+              .index_id = rdb_netbuf_to_uint32(
+                  reinterpret_cast<const uchar *>(txn.m_waiting_key.c_str()))};
+          std::string table_name = ddl_manager.safe_get_table_name(gl_index_id);
+          auto kd = ddl_manager.safe_find(gl_index_id);
+          std::string idx_name = kd->get_name();
+
+          rocksdb::ColumnFamilyHandle *cfh = cf_manager.get_cf(txn.m_cf_id);
+          std::string cf_name = cfh->GetName();
+
+          m_data += format_string("TRANSACTIONID: %u\n"
+                                  "COLUMN FAMILY NAME: %s\n"
+                                  "WAITING KEY: %s\n"
+                                  "LOCK TYPE: %s\n"
+                                  "INDEX NAME: %s\n"
+                                  "TABLE NAME: %s\n"
+                                  "---------------WAITING FOR---------------\n",
+                                  txn.m_txn_id, cf_name.c_str(),
+                                  rdb_hexdump(txn.m_waiting_key.c_str(),
+                                              txn.m_waiting_key.length(),
+                                              max_key_size)
+                                      .c_str(),
+                                  txn.m_exclusive ? "EXCLUSIVE" : "SHARED",
+                                  idx_name.c_str(), table_name.c_str());
+        }
+      } else {
+        m_data += "-------DEADLOCK EXCEEDED MAX DEPTH-------\n";
+      }
+
+      if (!path_entry.path.empty()) {
+        if (!path_entry.limit_exceeded) {
+          auto txn = path_entry.path[0];
```

Comment:
Since this code is duplicated, I wonder if it'd be cleaner if you wrote a helper for this.